src/compiler/checker.ts

```diff
@@ -11069,66 +11127,145 @@ namespace ts {
             }
 
-            if (correspondingPropType) {
-                checkTypeAssignableTo(exprType, correspondingPropType, node);
-            }
+            return attributesArray;
+        }
 
-            nameTable[node.name.text] = true;
-            return exprType;
+        /**
+         * Create anonymous type from given attributes symbol table.
+         * @param jsxAttributesSymb a JsxAttributes node containing attributes in attributesTable
+         * @param attributesTable a symbol table of attributes property
+         */
+        function createJsxAttributesType(jsxAttributesSymb: Symbol, attributesTable: Map<Symbol>) {
+            const result = createAnonymousType(jsxAttributesSymb, attributesTable, emptyArray, emptyArray, /*stringIndexInfo*/ undefined, /*numberIndexInfo*/ undefined);
+            const freshObjectLiteralFlag = compilerOptions.suppressExcessPropertyErrors ? 0 : TypeFlags.FreshLiteral;
+            result.flags |= TypeFlags.JsxAttributes | TypeFlags.ContainsObjectLiteral | freshObjectLiteralFlag;
+            result.objectFlags |= ObjectFlags.ObjectLiteral;
+            return result;
         }
 
-        function checkJsxSpreadAttribute(node: JsxSpreadAttribute, elementAttributesType: Type, nameTable: Map<boolean>) {
-            const type = checkExpression(node.expression);
-            const props = getPropertiesOfType(type);
-            for (const prop of props) {
-                // Is there a corresponding property in the element attributes type? Skip checking of properties
-                // that have already been assigned to, as these are not actually pushed into the resulting type
-                if (!nameTable[prop.name]) {
-                    const targetPropSym = getPropertyOfType(elementAttributesType, prop.name);
-                    if (targetPropSym) {
-                        const msg = chainDiagnosticMessages(undefined, Diagnostics.Property_0_of_JSX_spread_attribute_is_not_assignable_to_target_property, prop.name);
-                        checkTypeAssignableTo(getTypeOfSymbol(prop), getTypeOfSymbol(targetPropSym), node, undefined, msg);
+        /**
+         * Check JSXAttributes. This function is used when we are trying to figure out call signature for JSX opening-like element.
+         * In "checkApplicableSignatureForJsxOpeningLikeElement", we get type of arguments by checking the JSX opening-like element attributes property with contextual type.
+         * @param node a JSXAttributes to be resolved of its typea
+         */
+        function checkJsxAttributes(node: JsxAttributes) {
+            const symbolArray = getJsxAttributesSymbolArrayFromAttributesProperty(node.parent as JsxOpeningLikeElement);
+            let argAttributesType = anyType as Type;
+            if (symbolArray) {
+                const symbolTable = createMap<Symbol>();
+                forEach(symbolArray, (attr) => {
+                    symbolTable[attr.name] = attr;

```

Comment:
what about hyphenated names? The other caller handles hyphenated names differently.